### PR TITLE
waveterm: init at 0.8.7

### LINKED ISSUE #280566: Package request: waveterm
waveterm

Modern open-source terminal emulator supporting inline rendering.

* homepage URL: https://www.waveterm.dev/
* source URL: https://github.com/wavetermdev/waveterm
* license: Apache-2.0
* platforms: linux, darwin

### PR DESCRIPTION
**Pull request by ghost**
- Init waveterm at 0.8.7, An open-source, cross-platform terminal for seamless workflows, https://www.waveterm.dev
- Closes #330918
- Closes #280566

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by JohnRTitor — COMMENTED** [review]
I'd try to get the darwin builds and aarch64 builds as well. Examples in #330918

**Comment by JohnRTitor** [comment]
Thanks, `aarch64`/`arm64` could be added as well?

**Comment by aucub** [comment]
> Thanks, `aarch64`/`arm64` could be added as well?

I will try to add linux-aarch64, but I am unable to test it on the device.

**Comment by JohnRTitor** [comment]
Sure, it's just getting the binary tarball, should just work as long as it builds I believe, I can try to run it with `binfmt` emulation.

File: pkgs/by-name/wa/waveterm/package.nix [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    installPhase = ''
+      runHook preInstall
+      mkdir -p $out/${pname} $out/bin
+      cp -r ./* $out/${pname}/
+      runHook postInstall
+    '';
+
+    preFixup = ''
+      makeWrapper $out/${pname}/${pname} $out/bin/${pname} \
```
Comment (JohnRTitor) [inline pkgs/by-name/wa/waveterm/package.nix:157]: Don't use pname, here and above, it'll break the whole derivation, if someone just overrides pname.

File: pkgs/by-name/wa/waveterm/package.nix [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        x86_64-darwin = "sha256-tsqw597gQIMnQ/OPZhrWwaRliF94KyS+ryajttDLqBA=";
+        aarch64-darwin = "sha256-PD38UBSNKuv836P/py/CtrLOlddHa0+w7R20YVY4Ybc=";
+      }
+      .${system} or throwSystem;
+  };
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "waveterm";
+      exec = pname + " --no-sandbox %U";
```
Comment (JohnRTitor) [inline pkgs/by-name/wa/waveterm/package.nix:65]:
```suggestion
      exec = "waveterm --no-sandbox %U";
```

executable name, unlikely to change, and using pname here is risky

File: pkgs/by-name/wa/waveterm/package.nix [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+    installPhase = ''
+      runHook preInstall
+      mkdir -p $out/Applications/Wave.app
+      cp -R . $out/Applications/Wave.app
+      runHook postInstall
+    '';
+  };
+in
+if stdenv.isDarwin then darwin else linux
```
Comment (JohnRTitor) [inline pkgs/by-name/wa/waveterm/package.nix:189]:
```suggestion
if stdenv.hostPlatform.isDarwin then darwin else linux
```

**Review by JohnRTitor — COMMENTED** [review]
Overall looks good except for the few nits above. I'll try a build once those are fixed. 

For darwin CC @DontEatOreo for checking if it builds but I am afraid to use the package on Darwin we might need to do https://github.com/NixOS/nixpkgs/pull/330918#issuecomment-2266051274

Note: as of now all CI passes (darwin and linux)

File: pkgs/by-name/wa/waveterm/package.nix [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    description = "An open-source, cross-platform terminal for seamless workflows";
+    homepage = "https://www.waveterm.dev";
+    license = lib.licenses.asl20;
+    platforms = [
+      "aarch64-linux"
+      "aarch64-darwin"
+      "x86_64-linux"
+      "x86_64-darwin"
+    ];
+    maintainers = with lib.maintainers; [ aucub ];
```
Comment (SuperSandro2000) [inline pkgs/by-name/wa/waveterm/package.nix:103]: we are missing sourceProvenance here

Comment (JohnRTitor) [reply pkgs/by-name/wa/waveterm/package.nix:103]: Yup, this should work.

```nix
sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
```

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 345714` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>waveterm</li>
  </ul>
</details>

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 345714` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 345714` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Review by JohnRTitor — APPROVED** [review]
Let's get this done, I assume darwin too is fine since the CI is green.

**Comment by DontEatOreo** [comment]
Sorry for the late reply, but I have same question as I did in https://github.com/NixOS/nixpkgs/pull/330918#issuecomment-2266052612

**Comment by JohnRTitor** [comment]
Indeed we should, but I did not want to block this PR.

@DontEatOreo Another thing, if 7zz [corrupts](https://github.com/NixOS/nixpkgs/pull/330918#issuecomment-2263667006) the macos package, can we just use hdiutil, Official one is closed source, but there is a third party package that implements some functionality (`attach` and `detach` is all we need) 
https://github.com/go-darwin/hdiutil . It is pretty old, could you investigate if this still works?

**Comment by DontEatOreo** [comment]
I will take a look. I think maybe a hacky alternative could be to just have a Bash or Python script run `xattr -d com.apple.quarantine` on it, which would remove the warning

**Pull request merged**